deps/chakrashim/src/jsrtisolateshim.cc

```diff
@@ -670,6 +679,40 @@ void IsolateShim::SetPromiseRejectCallback(v8::PromiseRejectCallback callback) {
                                    reinterpret_cast<void*>(callback));
 }
 
+void IsolateShim::RunMicrotasks() {
+  // Loop until we've handled all the tasks, including the ones added by these tasks
+  while(this->microtaskQueue != nullptr) {
+    MicroTask* batch = this->microtaskQueue;
+    this->microtaskQueue = nullptr;
+
+    // Ensure that we handle things in-order
+    MicroTask* ordered = nullptr;
+    for(MicroTask* t = batch; t != nullptr;) {
+      MicroTask* next = t->next;
+      t->next = ordered;
+      ordered = t;
+      t = next;
+    }
+    while(ordered != nullptr) {;
+      MicroTask* mt = ordered;
+      ordered = ordered->next;
+      JsValueRef task = mt->task;
+      delete mt;
+
+      JsValueRef notUsed;
+      if (jsrt::CallFunction(task, &notUsed) != JsNoError) {
+        JsGetAndClearException(&notUsed);  // swallow any exception from task
+      }
+      JsRelease(task, nullptr);
+    }
+  }
+}
+
+void IsolateShim::QueueMicrotask(JsValueRef task) {
+  JsAddRef(task, nullptr);
+  this->microtaskQueue = new MicroTask(task, this->microtaskQueue);
```

Comment:
What if this fails? Do we leak the task?

Reply:
By "this" do you mean the `new` allocation? If that fails, then things are probably in a bad way (this is only allocating 2 pointers worth of data) but I hadn't given much thought to that scenario. The previous behavior was that we would push the task into an array in javascript, which I guess would OOM in the worst case and terminate the process, while here if the allocation fails then we leak any prior tasks, and also never run them.